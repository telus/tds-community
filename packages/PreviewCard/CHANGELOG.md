# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [1.3.0](https://github.com/telus/tds-community/compare/@tds/community-preview-card@1.2.0...@tds/community-preview-card@1.3.0) (2021-09-21)


### Features

* **community-preview-card:** react 17 and styled components 5 ([47ceaff](https://github.com/telus/tds-community/commit/47ceaffe00df420e70ecb3ee80a604708ea0fc6c))





# [1.2.0](https://github.com/telus/tds-community/compare/@tds/community-preview-card@1.1.7...@tds/community-preview-card@1.2.0) (2021-02-04)


### Features

* **community-preview-card:** add support for videos ([3d0f6bd](https://github.com/telus/tds-community/commit/3d0f6bdfd2c6dd6744edaf74e0a2e79ab03d97b7))





## [1.1.7](https://github.com/telus/tds-community/compare/@tds/community-preview-card@1.1.6...@tds/community-preview-card@1.1.7) (2020-09-09)

**Note:** Version bump only for package @tds/community-preview-card





## [1.1.6](https://github.com/telus/tds-community/compare/@tds/community-preview-card@1.1.5...@tds/community-preview-card@1.1.6) (2020-09-09)


### Bug Fixes

* **community-preview-card:** update util-helpers version ([35205b1](https://github.com/telus/tds-community/commit/35205b11cd77a13339403f74cbc7809074fc2ff7))





## [1.1.5](https://github.com/telus/tds-community/compare/@tds/community-preview-card@1.1.4...@tds/community-preview-card@1.1.5) (2020-07-31)

**Note:** Version bump only for package @tds/community-preview-card





## [1.1.4](https://github.com/telus/tds-community/compare/@tds/community-preview-card@1.1.3...@tds/community-preview-card@1.1.4) (2020-04-13)

**Note:** Version bump only for package @tds/community-preview-card





## [1.1.3](https://github.com/telus/tds-community/compare/@tds/community-preview-card@1.1.2...@tds/community-preview-card@1.1.3) (2020-03-19)

**Note:** Version bump only for package @tds/community-preview-card





## [1.1.2](https://github.com/telus/tds-community/compare/@tds/community-preview-card@1.1.1...@tds/community-preview-card@1.1.2) (2019-09-17)

**Note:** Version bump only for package @tds/community-preview-card





## [1.1.1](https://github.com/telus/tds-community/compare/@tds/community-preview-card@1.1.0...@tds/community-preview-card@1.1.1) (2019-06-13)

**Note:** Version bump only for package @tds/community-preview-card





# [1.1.0](https://github.com/telus/tds-community/compare/@tds/community-preview-card@1.0.1...@tds/community-preview-card@1.1.0) (2019-05-24)


### Features

* **community-preview-card:** add default props and make props unrequired ([a61e928](https://github.com/telus/tds-community/commit/a61e928))





## [1.0.1](https://github.com/telus/tds-community/compare/@tds/community-preview-card@1.0.0...@tds/community-preview-card@1.0.1) (2019-05-17)

**Note:** Version bump only for package @tds/community-preview-card





# 1.0.0 (2019-05-13)


### Bug Fixes

* **community-preview-card:** fix issue where card creates margin ([294df6f](https://github.com/telus/tds-community/commit/294df6f))
* **community-preview-card:** preview card takes component prop type ([4e69ad9](https://github.com/telus/tds-community/commit/4e69ad9))
* **community-preview-card:** removed width for responsiveness ([febba62](https://github.com/telus/tds-community/commit/febba62))


### Features

* **community-preview-card:** added the preview card component ([091bcd1](https://github.com/telus/tds-community/commit/091bcd1))
