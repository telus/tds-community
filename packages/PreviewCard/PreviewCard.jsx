import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import safeRest from '@tds/shared-safe-rest'
import Box from '@tds/core-box'
import Text from '@tds/core-text'
import { componentWithName } from '@tds/util-prop-types'
import { colorTelusPurple, colorWhiteLilac } from '@tds/core-colours'

/**
 * The PreviewCard component creates the appearance of a page snippet, and can be used in a list format.
 * @version ./package.json
 */
const PreviewCard = ({ category, other, image, body, footer, ...rest }) => {
  let header = category || other
  if (category && other) {
    header = `${category} \u00B7 ${other}`
  }

  let newBody = body
  if (body.length > 70) {
    newBody = `${body.substr(0, 70)}...`
  }

  const Anchor = styled.a`
    text-decoration: none;
    cursor: pointer;
  `

  const BoxContainer = styled.div`
    opacity: 1;
    border: 1px solid ${colorWhiteLilac};
    border-radius: 4px;
    width: ${header || footer ? '100%' : '269px'};
    max-width: 346px;
    height: ${header || footer ? '465px' : '305px'};
    @media (max-width: 576px) {
      height: auto;
      width: 100%;
    }
    overflow: hidden;
    &:hover {
      box-shadow: 0 0 16px 0 rgba(213, 213, 213, 0.5);
    }
  `

  const ContentContainer = styled.div`
    line-height: ${header && footer && '21px'};
  `

  const ImageContainer = styled.div`
    border-radius: 4px 4px 0 0;
    width: ${header || footer ? '100%' : '268px'};
    overflow: hidden;
    line-height: 0px;
    & > img {
      transition: all 0.3s ease-in-out;
      &:hover {
        transform: scale(1.03);
      }
    }
  `

  const P = styled.p`
    color: ${colorTelusPurple};
  `

  return (
    <Anchor {...safeRest(rest)}>
      <BoxContainer>
        {image && <ImageContainer>{image}</ImageContainer>}
        <Box horizontal={header || footer ? 4 : 3} vertical={header || footer ? 5 : 3}>
          <ContentContainer id="contentContainer">
            {header && (
              <Box>
                <Text size="small" bold>
                  {header}
                </Text>
              </Box>
            )}
            <Box vertical={header && 3}>
              <P alt={body}>{newBody}</P>
            </Box>
            {footer && (
              <Box>
                <Text size="small">{footer}</Text>
              </Box>
            )}
          </ContentContainer>
        </Box>
      </BoxContainer>
    </Anchor>
  )
}

PreviewCard.propTypes = {
  /**
   * Image component that will appear at the top of the card, above the content section.  Recommended dimensions is 369x269px.
   */
  image: componentWithName('Image').isRequired,
  /**
   * Text that will appear at the top of the content section.  Recommended to be only one or two words.
   */
  category: PropTypes.string,
  /**
   *  Text that will appear at the top of the content section, next to category seperated by a dot. Recommended to be 3 or less words.
   */
  other: PropTypes.string,
  /**
   * Purple text that will appear in the middle of the content section.  Recommended amount of characters is 70 or less.
   */
  body: PropTypes.string.isRequired,
  /**
   * Text that will appear at the bottom of the content section.  Recommended to be 3 or less words.
   */
  footer: PropTypes.string,
  /**
   * Target URL.
   */
  href: PropTypes.string.isRequired,
}

PreviewCard.defaultProps = {
  category: undefined,
  other: undefined,
  footer: undefined,
}

export default PreviewCard
