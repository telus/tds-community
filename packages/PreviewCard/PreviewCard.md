### Usage Criteria

- Use when you want to showcase and guide your users to a webpage.
- Using an image is a design requirement and is strongly recommended.

```jsx
<FlexGrid limitWidth={false} gutter={true}>
  <FlexGrid.Row>
    <FlexGrid.Col xs={12} md={6} lg={4}>
      <PreviewCard
        header="Data Intelligence \u00B7 March 21, 2019"
        image={<img src="blog-example.jpg" alt="Image of co-workers collaborating" width="100%" />}
        body="Hello world, this preview card has all the props and has text that is over 70 characters in length"
        footer="By Emelyn Ticong"
        href="#"
      />
    </FlexGrid.Col>
    <FlexGrid.Col xs={12} md={6} lg={4}>
      <PreviewCard
        header="Data Intelligence"
        image={<img src="blog-example.jpg" alt="Image of co-workers collaborating" width="100%" />}
        body="Hello world, this preview card has a category"
        href="#"
      />
    </FlexGrid.Col>
    <FlexGrid.Col xs={12} md={6} lg={4}>
      <PreviewCard
        image={<img src="blog-example.jpg" alt="Image of co-workers collaborating" width="100%" />}
        body="Hello world, this preview card has no category nor footer"
        href="#"
      />
    </FlexGrid.Col>
  </FlexGrid.Row>
</FlexGrid>
```
