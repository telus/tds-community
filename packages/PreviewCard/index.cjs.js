require('./dist/index.css')
const PreviewCard = require('./dist/index.cjs')

module.exports = PreviewCard
