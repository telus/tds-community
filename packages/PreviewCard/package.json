{
  "name": "@tds/community-preview-card",
  "version": "1.0.0-preview",
  "description": "",
  "main": "index.cjs.js",
  "module": "index.es.js",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/telus/tds-community.git"
  },
  "publishConfig": {
    "access": "public"
  },
  "author": "TELUS digital",
  "engines": {
    "node": ">=8"
  },
  "bugs": {
    "url": "https://github.com/telus/tds-community/issues"
  },
  "homepage": "http://tds.telus.com",
  "peerDependencies": {
    "react": ">=15",
    "react-dom": ">=15",
    "styled-components": "^4.2.0"
  },
  "dependencies": {
    "@tds/core-colours": "^1.1.1",
    "jest-styled-components": "^6.3.1",
    "prop-types": "^15.6.2"
  },
  "devDependencies": {
    "@tds/shared-safe-rest": "^1.0.0"
  }
}
