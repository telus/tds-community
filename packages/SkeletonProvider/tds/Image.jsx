import React from 'react'
import PropTypes from 'prop-types'
import Image from '@tds/core-image'
import withSkeleton from '../withSkeleton'
import styles from '../Skeleton.scss'

const ImageSkeleton = ({ skeleton, ...rest }) => {
  if (!skeleton) {
    return <Image {...rest} />
  }

  const baseClass = styles['skeleton-image']
  const circularClass = styles.circular

  const cssWidth = `${rest.width * 0.8}px`
  const cssHeight = `${rest.height * 0.8}px`

  const classes = [baseClass, circularClass].join(' ')

  return <div className={classes} style={{ width: cssWidth, height: cssHeight }} />
}

ImageSkeleton.propTypes = {
  skeleton: PropTypes.bool,
}

ImageSkeleton.defaultProps = {
  skeleton: false,
}

export default withSkeleton(ImageSkeleton)(Image)
