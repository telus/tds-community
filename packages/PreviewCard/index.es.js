import './dist/index.css'
import PreviewCard from './dist/index.es'

export default PreviewCard
