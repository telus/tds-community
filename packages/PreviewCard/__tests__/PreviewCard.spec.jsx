import React from 'react'
import { shallow, mount } from 'enzyme'
import 'jest-styled-components'

import Image from '@tds/core-image'

import PreviewCard from '../PreviewCard'

describe('PreviewCard', () => {
  const doShallow = (props = {}) => shallow(<PreviewCard {...props} />)

  const requiredProps = {
    body: 'Hello world!',
    image: (
      <Image
        src="blog-example.jpg"
        alt="Image of co-workers collaborating"
        width={300}
        height={300}
      />
    ),
    href: '#',
  }

  it('renders with only Required props', () => {
    const previewCard = doShallow({ ...requiredProps })

    expect(previewCard).toMatchSnapshot()
  })

  it('renders with category and other', () => {
    const previewCard = doShallow({
      ...requiredProps,
      category: 'development',
      other: 'April 1st, 2019',
    })

    expect(previewCard).toMatchSnapshot()
  })

  it('renders with footer', () => {
    const previewCard = doShallow({ ...requiredProps, footer: 'By Halo' })

    expect(previewCard).toMatchSnapshot()
  })

  it('renders with all props', () => {
    const previewCard = doShallow({
      ...requiredProps,
      category: 'development',
      other: 'April 1st, 2019',
      footer: 'By Halo',
    })

    expect(previewCard).toMatchSnapshot()
  })

  it('renders with long body text', () => {
    const previewCard = shallow(
      <PreviewCard
        image={
          <Image
            src="blog-example.jpg"
            alt="Image of co-workers collaborating"
            width={300}
            height={300}
          />
        }
        body="Hello world, this preview card has all the props and has text that is over 70 characters in length"
        href="#"
      />
    )

    expect(previewCard).toMatchSnapshot()
  })

  it('passes additional attributes to the element', () => {
    const previewCard = doShallow({
      ...requiredProps,
      id: 'the-id',
      'data-some-attr': 'some value',
    })

    expect(previewCard).toHaveProp('id', 'the-id')
    expect(previewCard).toHaveProp('data-some-attr', 'some value')
  })

  it('does not allow custom CSS', () => {
    const previewCard = doShallow({
      ...requiredProps,
      className: 'my-custom-class',
      style: { color: 'hotpink' },
    })

    expect(previewCard).not.toHaveProp('className', 'my-custom-class')
    expect(previewCard).not.toHaveProp('style')
  })

  it('renders with line height of 21px', () => {
    const previewCard = mount(
      <PreviewCard
        category="development"
        other="April 1st, 2019"
        footer="By Halo"
        {...requiredProps}
      />
    ).find('[data-testid="contentContainer"]')

    expect(previewCard).toHaveStyleRule('line-height', '21px')
  })
})
