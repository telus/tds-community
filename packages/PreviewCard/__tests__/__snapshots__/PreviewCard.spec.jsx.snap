// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreviewCard renders with all props 1`] = `
<styled.a
  href="#"
>
  <styled.div
    isLarge="development · April 1st, 2019"
  >
    <styled.div
      isLarge="development · April 1st, 2019"
    >
      <Image
        alt="Image of co-workers collaborating"
        height={300}
        src="blog-example.jpg"
        width={300}
      />
    </styled.div>
    <Box
      horizontal={4}
      inline={false}
      tag="div"
      vertical={5}
    >
      <styled.div
        data-testid="contentContainer"
        footer="By Halo"
        header="development · April 1st, 2019"
      >
        <Box
          inline={false}
          tag="div"
        >
          <Text
            block={false}
            bold={true}
            invert={false}
            size="small"
          >
            development · April 1st, 2019
          </Text>
        </Box>
        <Box
          inline={false}
          tag="div"
          vertical={3}
        >
          <styled.p
            alt="Hello world!"
          >
            Hello world!
          </styled.p>
        </Box>
        <Box
          inline={false}
          tag="div"
        >
          <Text
            block={false}
            bold={false}
            invert={false}
            size="small"
          >
            By Halo
          </Text>
        </Box>
      </styled.div>
    </Box>
  </styled.div>
</styled.a>
`;

exports[`PreviewCard renders with footer 1`] = `
<styled.a
  href="#"
>
  <styled.div
    isLarge="By Halo"
  >
    <styled.div
      isLarge="By Halo"
    >
      <Image
        alt="Image of co-workers collaborating"
        height={300}
        src="blog-example.jpg"
        width={300}
      />
    </styled.div>
    <Box
      horizontal={4}
      inline={false}
      tag="div"
      vertical={5}
    >
      <styled.div
        data-testid="contentContainer"
        footer="By Halo"
      >
        <Box
          inline={false}
          tag="div"
          vertical={3}
        >
          <styled.p
            alt="Hello world!"
          >
            Hello world!
          </styled.p>
        </Box>
        <Box
          inline={false}
          tag="div"
        >
          <Text
            block={false}
            bold={false}
            invert={false}
            size="small"
          >
            By Halo
          </Text>
        </Box>
      </styled.div>
    </Box>
  </styled.div>
</styled.a>
`;

exports[`PreviewCard renders with header 1`] = `
<styled.a
  href="#"
>
  <styled.div
    isLarge="development · April 1st, 2019"
  >
    <styled.div
      isLarge="development · April 1st, 2019"
    >
      <Image
        alt="Image of co-workers collaborating"
        height={300}
        src="blog-example.jpg"
        width={300}
      />
    </styled.div>
    <Box
      horizontal={4}
      inline={false}
      tag="div"
      vertical={5}
    >
      <styled.div
        data-testid="contentContainer"
        header="development · April 1st, 2019"
      >
        <Box
          inline={false}
          tag="div"
        >
          <Text
            block={false}
            bold={true}
            invert={false}
            size="small"
          >
            development · April 1st, 2019
          </Text>
        </Box>
        <Box
          inline={false}
          tag="div"
          vertical={3}
        >
          <styled.p
            alt="Hello world!"
          >
            Hello world!
          </styled.p>
        </Box>
      </styled.div>
    </Box>
  </styled.div>
</styled.a>
`;

exports[`PreviewCard renders with long body text 1`] = `
<styled.a
  href="#"
>
  <styled.div>
    <styled.div>
      <Image
        alt="Image of co-workers collaborating"
        height={300}
        src="blog-example.jpg"
        width={300}
      />
    </styled.div>
    <Box
      horizontal={3}
      inline={false}
      tag="div"
      vertical={3}
    >
      <styled.div
        data-testid="contentContainer"
      >
        <Box
          inline={false}
          tag="div"
        >
          <styled.p
            alt="Hello world, this preview card has all the props and has text that is over 70 characters in length"
          >
            Hello world, this preview card has all the props and has text that is ...
          </styled.p>
        </Box>
      </styled.div>
    </Box>
  </styled.div>
</styled.a>
`;

exports[`PreviewCard renders with only Required props 1`] = `
<styled.a
  href="#"
>
  <styled.div>
    <styled.div>
      <Image
        alt="Image of co-workers collaborating"
        height={300}
        src="blog-example.jpg"
        width={300}
      />
    </styled.div>
    <Box
      horizontal={3}
      inline={false}
      tag="div"
      vertical={3}
    >
      <styled.div
        data-testid="contentContainer"
      >
        <Box
          inline={false}
          tag="div"
        >
          <styled.p
            alt="Hello world!"
          >
            Hello world!
          </styled.p>
        </Box>
      </styled.div>
    </Box>
  </styled.div>
</styled.a>
`;
